src/coreclr/jit/morph.cpp

```diff
@@ -15609,6 +15547,99 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
     return tree;
 }
 
+//------------------------------------------------------------------------
+// fgOptimizeRelationalComparisonWithConst: optimizes a comparison operation.
+//
+// Recognizes comparisons against various constant operands and morphs
+// them, if possible, into comparisons against zero.
+//
+// Arguments:
+//   cmp - the GT_LE/GT_LT/GT_GE/GT_GT tree to morph.
+//
+// Return Value:
+//   The "cmp" tree, possibly with a modified oper.
+//   The second operand's constant value may be modified as well.
+//
+// Assumptions:
+//   The operands have been swapped so that any constants are on the right.
+//   The second operand is an integral constant.
+//
+GenTree* Compiler::fgOptimizeRelationalComparisonWithConst(GenTreeOp* cmp)
+{
+    assert(cmp->OperIs(GT_LE, GT_LT, GT_GE, GT_GT));
+    assert(cmp->gtGetOp2()->OperIs(GT_CNS_INT));
+
+    GenTree*             op1 = cmp->gtGetOp1();
+    GenTreeIntConCommon* op2 = cmp->gtGetOp2()->AsIntConCommon();
+
+    assert(genActualType(op1) == genActualType(op2));
+
+    genTreeOps oper     = cmp->OperGet();
+    int64_t    op2Value = op2->IntegralValue();
+
+    if (op2Value == 1)
+    {
+        // Check for "expr >= 1".
+        if (oper == GT_GE)
+        {
+            // Change to "expr != 0" for unsigned and "expr > 0" for signed.
+            oper = cmp->IsUnsigned() ? GT_NE : GT_GT;
+        }
+        // Check for "expr < 1".
+        else if (oper == GT_LT)
+        {
+            // Change to "expr == 0" for unsigned and "expr <= 0".
+            oper = cmp->IsUnsigned() ? GT_EQ : GT_LE;
+        }
+    }
+    // Check for "expr relop -1".
+    else if (!cmp->IsUnsigned() && (op2Value == -1))
+    {
+        // Check for "expr <= -1".
+        if (oper == GT_LE)
+        {
+            // Change to "expr < 0".
+            oper = GT_LT;
+        }
+        // Check for "expr > -1".
+        else if (oper == GT_GT)
+        {
+            // Change to "expr >= 0".
+            oper = GT_GE;
+        }
+    }
+    else if (cmp->IsUnsigned())
+    {
+        if ((oper == GT_LE) || (oper == GT_GT))
+        {
+            if (op2Value == 0)
+            {
+                // IL doesn't have a cne instruction so compilers use cgt.un instead. The JIT
+                // recognizes certain patterns that involve GT_NE (e.g (x & 4) != 0) and fails
+                // if GT_GT is used instead. Transform (x GT_GT.unsigned 0) into (x GT_NE 0)
+                // and (x GT_LE.unsigned 0) into (x GT_EQ 0). The later case is rare, it sometimes
+                // occurs as a result of branch inversion.
+                oper = (oper == GT_LE) ? GT_EQ : GT_NE;
+                cmp->gtFlags &= ~GTF_UNSIGNED;
+            }
+        }
+    }
+
+    if (!cmp->OperIs(oper))
+    {
+        // Keep the old ValueNumber for 'tree' as the new expr
+        // will still compute the same value as before.
+        cmp->SetOper(oper, GenTree::PRESERVE_VN);
+        op2->SetIntegralValue(0);
```

Comment:
> it is the only use and you can use `SetIconValue(0)` because you have `assert(cmp->gtGetOp2()->OperIs(GT_CNS_INT));`

Hmm, I have `cmp->gtGetOp2()->IsIntegralConst()`, which passes for `GT_CNS_LNG` as well, and `SetIconValue` has an `assert(gtOper == GT_CNS_INT);`, so I think `SetIntegralValue` is still needed.

Reply:
Ah, I was reviewing by commits and the condition was changed from `if (op2->OperIs(GT_CNS_INT))` 
https://github.com/dotnet/runtime/pull/54204/commits/e96baef6320e59d2ae571fd7039042c42a7ada9c  to `op2->IsIntegralConst()` https://github.com/dotnet/runtime/pull/54204/commits/bf4c49f09d558247876f9eec6f56aa8f95aab210 .

Now it is clear, thanks.